File: src/Web/Dart/state.dart
enum GateState { locked, unlocked }

enum GateAction { insertCoin, push }

({GateState state, String result}) transition(
  GateState state,
  GateAction action,
) => switch ((state, action)) {
  (GateState.locked, GateAction.insertCoin) => (
    state: GateState.unlocked,
    result: 'unlocked',
  ),
  (GateState.unlocked, GateAction.push) => (
    state: GateState.locked,
    result: 'passed',
  ),
  (GateState.locked, GateAction.push) => (
    state: GateState.locked,
    result: 'blocked',
  ),
  (GateState.unlocked, GateAction.insertCoin) => (
    state: GateState.unlocked,
    result: 'already-unlocked',
  ),
};

void check(bool condition, String message) {
  if (!condition) throw StateError(message);
}

void main() {
  var state = GateState.locked;
  check(state == GateState.locked, 'turnstile must start locked');

  var outcome = transition(state, GateAction.push);
  check(outcome.state == GateState.locked, 'invalid push must preserve locked');
  check(outcome.result == 'blocked', 'locked push must be blocked');

  outcome = transition(outcome.state, GateAction.insertCoin);
  check(outcome.state == GateState.unlocked, 'coin must unlock turnstile');
  check(outcome.result == 'unlocked', 'unlock result must be observable');

  outcome = transition(outcome.state, GateAction.insertCoin);
  check(
    outcome.state == GateState.unlocked,
    'duplicate coin must preserve unlocked',
  );
  check(
    outcome.result == 'already-unlocked',
    'duplicate coin behavior must depend on current state',
  );

  outcome = transition(outcome.state, GateAction.push);
  check(outcome.state == GateState.locked, 'valid push must return to locked');
  check(outcome.result == 'passed', 'unlocked push must pass');

  print('dart-state: passed');
}

File: src/Systems/Zig/state.zig
const std = @import("std");

const GateState = enum {
    locked,
    unlocked,
};

const GateAction = enum {
    lock,
    unlock,
};

fn transition(state: GateState, action: GateAction) GateState {
    return switch (state) {
        .locked => if (action == .unlock) .unlocked else .locked,
        .unlocked => if (action == .lock) .locked else .unlocked,
    };
}

pub fn runContract() bool {
    const initial: GateState = .locked;
    if (transition(initial, .lock) != .locked) return false;

    const unlocked = transition(initial, .unlock);
    if (unlocked != .unlocked) return false;
    if (transition(unlocked, .unlock) != .unlocked) return false;

    return transition(unlocked, .lock) == .locked;
}

pub fn main() !void {
    if (!runContract()) return error.StateContractFailed;
    std.debug.print("zig-state: passed\n", .{});
}
